### PR TITLE
Disable validation completely when `overfit_batches>0` 

### LINKED ISSUE #8962: `limit_val_batches=0.0` has no effect when `overfit_batches=1`
## 🐛 Bug

I want to overfit a batch using `overfit_batches`, `limit_val_batches` is set to `0.0` but there's still a validation batch every epoch making it take twice as long. Why is validation happening at all when using `overfit_batches` and why is `limit_val_batches` not working as expected?

This is my trainer, 
```
trainer = pl.Trainer(
    gpus=1,
    overfit_batches=1,
    limit_val_batches=0.0,
    log_every_n_steps=1,
    deterministic=True,
)
```

### To Reproduce

I tried to reproduce using the boring model, but the colab nb is broken 
image

### Expected behavior

Validation shouldn't happen by default when using overfit_batches. If it is desirable to have a validation batch then `limit_val_batches=0.0` should override that behavior. 

### Environment

pytorch 1.9.0
lightning 1.4.1

### LINKED ISSUE #9989: Remove validation completely when overfit_batches > 0
## Proposed refactoring or deprecation

<details><summary>CODE</summary>
<p>

```py
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch['a']).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        breakpoint()
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

if __name__ == '__main__':
    train_data = {'a': DataLoader(RandomDataset(32, 64), batch_size=2), 'b': DataLoader(RandomDataset(32, 64), batch_size=2)}
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)


    model = BoringModel()
    trainer = Trainer(
        gpus=None,
        max_epochs=2,
        num_sanity_val_steps=0,
        overfit_batches=2
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
```

</p>
</details>

I think we should just turn of validation when `overfit_batches > 0` instead of silently replacing the `val_dataloaders` with `train_dataloader` here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/8c76cf5ae1db078dcd201b08de788ac6769c894d/pytorch_lightning/trainer/data_loading.py#L414-L416

### Motivation

There is a possibility that the structure of data returned by train dataloader might not be compatible with `validation_step` (check example above) and thus will lead to errors. To make it compatible, users might have to update their `validation_step` code just for `overfit_batches > 0`, which I guess isn't a good user-exp and also leads to errors. Overfit_batches is just used for quick debugging and checking model adaptability with the train dataset.

Also the above code leads to error even before `validation_step` is called since train_dataloaders are not handled the right way.

### Pitch

Turn off validation completely by setting `limit_val_batches=0` during init if `overfit_batches > 0`.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @justusschock @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by popfido**
## What does this PR do?

Fixes #9989 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
@rohitgr7 Any updates on this ?

**Comment by rohitgr7** [comment]
@popfido 
let us know if you need any help.
Also, I'd just update this: https://github.com/PyTorchLightning/pytorch-lightning/blob/48cf1adfd3ad9c7e659083a4afc334dafb331f28/pytorch_lightning/trainer/trainer.py#L1772-L1776
and tests.

Also can you update the changelog and pr description.

**Comment by popfido** [comment]
@rohitgr7 PR description modified. I found that lots of test cases have used `overfit_batches` with lots of assertion on action of validation step. This may takes some time to deprecate. Is that okay?

I'll modify the changelog after all things done.

**Comment by rohitgr7** [comment]
> @rohitgr7 PR description modified. I found that lots of test cases have used `overfit_batches` with lots of assertion on action of validation step. This may takes some time to deprecate. Is that okay?
> 
> I'll modify the changelog after all things have been done.

totally fine! 😃

**Comment by popfido** [comment]
@rohitgr7 Some test cases was canceled by no reason, the other is okay. Ready for Review.

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -626,7 +626,7 @@ def _determine_data_use_amount(self, overfit_batches: float) -> None:
         """Use less data for debugging purposes."""
         if overfit_batches > 0:
             self.limit_train_batches = overfit_batches
-            self.limit_val_batches = overfit_batches
+            self.limit_val_batches = 0
             self.limit_test_batches = overfit_batches
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:630]:
```suggestion
```
we don't need to touch testing at all.

File: tests/callbacks/test_early_stopping.py [hunk 2]
```diff
@@ -228,7 +228,7 @@ def validation_epoch_end(self, outputs):
     early_stopping = EarlyStopping(
         monitor="abc", stopping_threshold=stopping_threshold, divergence_threshold=divergence_theshold
     )
-    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], overfit_batches=0.20, max_epochs=20)
+    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], max_epochs=20)
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:231]: I guess overfit_batches was used here to limit the batches for both train and val.
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], limit_train_batches=0.2, limit_val_batches=0.2, max_epochs=20)
```

File: tests/callbacks/test_early_stopping.py [hunk 3]
```diff
@@ -246,7 +246,7 @@ def validation_epoch_end(self, outputs):
 
     model = CurrentModel()
     early_stopping = EarlyStopping(monitor="val_loss", check_finite=True)
-    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], overfit_batches=0.20, max_epochs=10)
+    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], max_epochs=10)
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:249]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], limit_train_batches=0.2, limit_val_batches=0.2, max_epochs=10)
```

File: tests/callbacks/test_early_stopping.py [hunk 4]
```diff
@@ -426,7 +426,6 @@ def test_multiple_early_stopping_callbacks(
     trainer = Trainer(
         default_root_dir=tmpdir,
         callbacks=callbacks,
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:428]:
```suggestion
        callbacks=callbacks,
        limit_train_batches=0.2,
        limit_val_batches=0.2,
```

File: tests/models/test_cpu.py [hunk 5]
```diff
@@ -106,7 +106,6 @@ def validation_step(self, *args, **kwargs):
         callbacks=[stopping],
         default_root_dir=tmpdir,
         gradient_clip_val=1.0,
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:108]:
```suggestion
        gradient_clip_val=1.0,
        limit_train_batches=0.2,
        limit_val_batches=0.2,
```

File: tests/trainer/test_trainer_tricks.py [hunk 6]
```diff
@@ -112,33 +112,34 @@ def test_overfit_batch_limits(tmpdir):
     # ------------------------------------------------------
     for split in (RunningStage.VALIDATING, RunningStage.TESTING):
```
Comment (rohitgr7) [inline tests/trainer/test_trainer_tricks.py:113]: can you check for both of the stages here?
for val it should be 0 and for test it shouldn't change at all.

File: tests/trainer/flags/test_overfit_batches.py [hunk 7]
```diff
@@ -16,41 +16,47 @@
...
+from tests.base.model_template import EvalModelTemplate
 from tests.helpers.boring_model import BoringModel, RandomDataset
 
 
-def test_overfit_multiple_val_loaders(tmpdir):
-    """Tests that overfit batches works with multiple val dataloaders."""
-    val_dl_count = 2
-    overfit_batches = 3
+@pytest.mark.parametrize("overfit_batches", [1, 2, 0.1, 0.25, 1.0])
+def test_disable_validation_when_overfit_batches_larger_than_zero(tmpdir, overfit_batches):
```
Comment (rohitgr7) [inline tests/trainer/flags/test_overfit_batches.py:24]: I think we can remove this test completely. Testing within `tests/trainer/test_trainer_tricks.py:test_overfit_batch_limits` would be enough.

**Review by rohitgr7 — COMMENTED** [review]
looks good so far.. left some comments 😃

File: tests/trainer/flags/test_overfit_batches.py (hunk 7 above)
Comment (popfido) [reply tests/trainer/flags/test_overfit_batches.py]: Done

File: tests/callbacks/test_early_stopping.py [hunk 8]
```diff
@@ -228,7 +228,13 @@ def validation_epoch_end(self, outputs):
...
         monitor="abc", stopping_threshold=stopping_threshold, divergence_threshold=divergence_theshold
     )
-    trainer = Trainer(default_root_dir=tmpdir, callbacks=[early_stopping], overfit_batches=0.20, max_epochs=20)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        callbacks=[early_stopping],
+        limit_train_batches=0.2,
+        limit_val_batches=0.2,
+        max_epochs=20,
+    )
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:237]: not related to this PR. I guess we need to update some of the tests here. they run for 20 epochs .. might improve ci runtime a little bit.

**Review by rohitgr7 — APPROVED** [review]
LGTM!!

File: tests/callbacks/test_early_stopping.py (hunk 8 above)
Comment (tchaton) [reply tests/callbacks/test_early_stopping.py:237]: Yes, the early stopping one is quite consuming. Definitely worth optimizing, same for the fault-tolerant ones.

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -626,8 +626,7 @@ def _determine_data_use_amount(self, overfit_batches: float) -> None:
         """Use less data for debugging purposes."""
         if overfit_batches > 0:
             self.limit_train_batches = overfit_batches
-            self.limit_val_batches = overfit_batches
-            self.limit_test_batches = overfit_batches
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:630]: Neat we don't override the test value anymore.

File: tests/trainer/test_trainer_tricks.py [hunk 10]
```diff
@@ -112,33 +112,46 @@ def test_overfit_batch_limits(tmpdir):
...
         trainer = Trainer(overfit_batches=1)
         trainer._data_connector.attach_dataloaders(model)
         loader_num_batches, dataloaders = trainer._reset_eval_dataloader(split, model=model)
-        assert loader_num_batches[0] == 1
+        if split == RunningStage.VALIDATING:
+            assert loader_num_batches[0] == 0
+        else:
+            assert loader_num_batches[0] == len(test_loader)
+            # make sure we turned off shuffle for the user
+            assert isinstance(dataloaders[0].sampler, SequentialSampler)
```
Comment (tchaton) [inline tests/trainer/test_trainer_tricks.py:141]: @rohitgr7 quick question. How should we support overfit_batches with DDP? I think we would need to make sure the DistributedSampler shuffle value is set to False.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -626,8 +626,7 @@ def _determine_data_use_amount(self, overfit_batches: float) -> None:
         """Use less data for debugging purposes."""
         if overfit_batches > 0:
             self.limit_train_batches = overfit_batches
-            self.limit_val_batches = overfit_batches
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:629]: Does this mean the following?

```python
trainer = Trainer(overfit_batches=3)
trainer.fit(model)  # works as before
trainer.validate()  # now does nothing
trainer.test()  # now does nothing
```

If that's the case, I think the later calls should raise a warning as this is a huge potential footgun. And another pro for #10444

File: tests/trainer/test_trainer_tricks.py [hunk 12]
```diff
@@ -112,33 +112,46 @@ def test_overfit_batch_limits(tmpdir):
     # ------------------------------------------------------
     for split in (RunningStage.VALIDATING, RunningStage.TESTING):
 
+        # ------------------------------------------------------
+        # test overfit_batches action
+        # ------------------------------------------------------
+
```
Comment (carmocca) [inline tests/trainer/test_trainer_tricks.py:118]: Why was this added?

File: CHANGELOG.md [hunk 13]
```diff
@@ -57,7 +57,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Raised an error if the `batch_size` cannot be inferred from the current batch if it contained a string or was a custom batch object ([#10541](https://github.com/PyTorchLightning/pytorch-lightning/pull/10541))
 
 
--
+- Disable validation completely when `overfit_batches > 0` ([#9709](https://github.com/PyTorchLightning/pytorch-lightning/pull/9709))
```
Comment (carmocca) [inline CHANGELOG.md:60]: Can we also check whether docs need to be updated? They probably mention the old behaviour

**Comment by popfido** [comment]
@rohitgr7 One of the test cases on macos is still in timeout, other issues are resolved

File: CHANGELOG.md (hunk 13 above)
Comment (popfido) [reply CHANGELOG.md:60]: Solved

File: tests/trainer/test_trainer_tricks.py (hunk 12 above)
Comment (popfido) [reply tests/trainer/test_trainer_tricks.py:118]: The test cases below are all sub-test of it. Or I can remove it.

File: tests/trainer/test_trainer_tricks.py (hunk 6 above)
Comment (popfido) [reply tests/trainer/test_trainer_tricks.py:113]: Checked, other issues metioned below.

File: tests/callbacks/test_early_stopping.py (hunk 8 above)
Comment (popfido) [reply tests/callbacks/test_early_stopping.py:237]: Im macos test env, the test case `test_fast_forward_sampler_with_distributed_sampler` cost much more time than that (about 60s) which I think might also affect ci runtime.

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (popfido) [reply pytorch_lightning/trainer/trainer.py:629]: this mean the following
`
trainer = Trainer(overfit_batches=3)
trainer.fit(model)  # works as before
trainer.validate()  # now does nothing
trainer.test()  # works as the same with a Trainer without setting `overfit_batches=3` 
`

File: CHANGELOG.md (hunk 13 above)
Comment (popfido) [reply CHANGELOG.md:60]: If there is any extra info need to be stated. Left here?

Comment (popfido) [reply CHANGELOG.md:60]: I can add extra info on logic change of `limit_test_batches` when `overfit_batches > 0`

File: tests/trainer/test_trainer_tricks.py (hunk 10 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer_tricks.py:141]: @tchaton yes we do here: https://github.com/PyTorchLightning/pytorch-lightning/blob/3d6262b7a91215e72019d720e742e6261e1636dc/pytorch_lightning/trainer/data_loading.py#L187

I guess there is no test for it.. I'll add one in a follow-up with some cleanup.

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:629]: yep, that's a limitation we get for setting these flags inside the trainer.

**Comment by rohitgr7** [comment]
> @rohitgr7 One of the test cases on macos is still in timeout, other issues are resolved

@popfido don't worry about that.. happens sometimes.

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:640]: we can rework this in another issue so that overfit batches only applies to the fit loop.

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:640]: I agree it would be expected as it is right now

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: CHANGELOG.md (hunk 13 above)
Comment (awaelchli) [reply CHANGELOG.md:71]: There are several references to the val and test dataset in trainer.rst, debugging.rst, speed.rst and maybe otheres if you do a full search for "overfit_batches"

File: CHANGELOG.md [hunk 14]
```diff
@@ -68,6 +68,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Raised an error if the `batch_size` cannot be inferred from the current batch if it contained a string or was a custom batch object ([#10541](https://github.com/PyTorchLightning/pytorch-lightning/pull/10541))
 
 
+- Disable validation completely when `overfit_batches > 0` ([#9709](https://github.com/PyTorchLightning/pytorch-lightning/pull/9709))
```
Comment (awaelchli) [inline CHANGELOG.md:71]:
```suggestion
- The validation loop is now disabled when `overfit_batches > 0` is set in the Trainer ([#9709](https://github.com/PyTorchLightning/pytorch-lightning/pull/9709))
```

File: CHANGELOG.md (hunk 13 above)
Comment (rohitgr7) [reply CHANGELOG.md]: yeah. I'll be doing a follow-up to clean up the code a bit.
